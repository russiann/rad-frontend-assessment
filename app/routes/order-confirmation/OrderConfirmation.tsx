import React from "react";
import type { Route } from "./+types/OrderConfirmation";
import { useOrderConfirmation } from "./useOrderConfirmation";
import { ConfirmationCard } from "./partials/ConfirmationCard";

export function meta({}: Route.MetaArgs) {
  return [
    { title: "Order Confirmation - React Shopping" },
    { name: "description", content: "Your order has been confirmed" },
  ];
}

export default function OrderConfirmation() {
  const { actions } = useOrderConfirmation();

  return (
    <div className="container mx-auto px-4 py-8 max-w-2xl flex justify-center items-center">
      <ConfirmationCard 
        onContinueShopping={actions.continueShopping}
        onReturnToHome={actions.returnToHome}
      />
    </div>
  );
} 