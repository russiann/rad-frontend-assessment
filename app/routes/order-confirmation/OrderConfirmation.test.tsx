import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import OrderConfirmation from './OrderConfirmation';
import type { ReactNode } from 'react';

// Mock dependencies
const mockNavigate = vi.fn();

// Mock react-router
vi.mock('react-router', () => ({
  useNavigate: () => mockNavigate,
  useLocation: () => ({ pathname: '/order-confirmation' }),
  useParams: () => ({}),
  Link: ({ children, to }: { children: React.ReactNode; to: string }) => (
    <a href={to}>{children}</a>
  ),
}));

describe('OrderConfirmation Page', () => {
  let queryClient: QueryClient;
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
        },
      },
    });

    user = userEvent.setup();
    mockNavigate.mockClear();
  });

  afterEach(() => {
    queryClient.clear();
  });

  const createWrapper = () => ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  describe('Basic Rendering', () => {
    it('should render order confirmation page with success message', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      expect(screen.getByText(/order confirmed/i)).toBeInTheDocument();
      expect(screen.getByText(/thank you for your purchase/i)).toBeInTheDocument();
    });

    it('should show confirmation card with order details', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      expect(screen.getByText(/your order has been received/i)).toBeInTheDocument();
      expect(screen.getByText(/you will receive an email confirmation shortly/i)).toBeInTheDocument();
    });

    it('should display success icon or checkmark', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      // Look for success styling or confirmation text
      expect(screen.getByText('Order Confirmed!')).toBeInTheDocument();
    });
  });

  describe('Navigation Actions', () => {
    it('should have continue shopping button', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const continueShoppingButton = screen.getByRole('button', { name: /continue shopping/i });
      expect(continueShoppingButton).toBeInTheDocument();
    });

    it('should have return to home button', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const returnToHomeButton = screen.getByRole('button', { name: /return to home/i }) ||
                                screen.getByRole('button', { name: /back to home/i });
      expect(returnToHomeButton).toBeInTheDocument();
    });

    it('should navigate to home when continue shopping is clicked', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const continueShoppingButton = screen.getByRole('button', { name: /continue shopping/i });
      await user.click(continueShoppingButton);

      expect(mockNavigate).toHaveBeenCalledWith('/');
    });

    it('should navigate to home when return to home is clicked', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const returnToHomeButton = screen.getByRole('button', { name: /return to home/i }) ||
                                screen.getByRole('button', { name: /back to home/i });
      await user.click(returnToHomeButton);

      expect(mockNavigate).toHaveBeenCalledWith('/');
    });
  });

  describe('Page Layout', () => {
    it('should be centered and well-positioned', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const container = document.querySelector('.container');
      expect(container).toBeTruthy();
    });

    it('should have proper spacing and styling', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      // The confirmation card should be rendered
      expect(screen.getByText('Order Confirmed!')).toBeInTheDocument();
    });
  });

  describe('User Experience', () => {
    it('should provide clear next steps for the user', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      // Should have clear action buttons for next steps
      expect(screen.getByRole('button', { name: /continue shopping/i })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: /return to home/i }) ||
             screen.getByRole('button', { name: /back to home/i })).toBeInTheDocument();
    });

    it('should have informative confirmation message', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      // Should contain informative text about the order
      expect(screen.getAllByText(/order/i)).toBeTruthy();
      expect(screen.getByText(/confirmed/i)).toBeInTheDocument();
    });

    it('should indicate that user will receive email confirmation', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      expect(screen.getByText(/email/i)).toBeInTheDocument();
      expect(screen.getByText(/confirmation/i)).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('should have proper heading structure', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const heading = screen.getByRole('heading', { level: 1 }) ||
                     screen.getByRole('heading', { level: 2 });
      expect(heading).toBeInTheDocument();
    });

    it('should have accessible button labels', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <OrderConfirmation />
        </TestWrapper>
      );

      const buttons = screen.getAllByRole('button');
      buttons.forEach(button => {
        expect(button).toHaveAccessibleName();
      });
    });
  });
});