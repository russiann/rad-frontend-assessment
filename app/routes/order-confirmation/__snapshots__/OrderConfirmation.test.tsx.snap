// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderConfirmation > snapshots > renders order confirmation correctly 1`] = `
<div
  data-overlay-container="true"
>
  <div
    class="container mx-auto px-4 py-8 max-w-2xl flex justify-center items-center"
  >
    <div
      class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none max-w-md w-full"
      tabindex="-1"
    >
      <div
        class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words overflow-y-auto subpixel-antialiased py-12 px-8 text-center"
      >
        <div
          class="flex justify-center mb-6"
        >
          <div
            class="bg-success/10 rounded-full p-4"
          >
            <span />
          </div>
        </div>
        <h1
          class="text-3xl font-semibold mb-3"
        >
          Order Confirmed!
        </h1>
        <p
          class="text-default-600 mb-8"
        >
          Thank you for your purchase. Your order has been received and is being processed. You will receive an email confirmation shortly.
        </p>
        <div
          class="flex flex-col gap-4"
        >
          <button
            class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-6 min-w-24 h-12 text-medium gap-3 rounded-large [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary text-primary-foreground data-[hover=true]:opacity-hover w-full"
            data-react-aria-pressable="true"
            tabindex="0"
            type="button"
          >
            <span />
            Continue Shopping
          </button>
          <button
            class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-6 min-w-24 h-12 text-medium gap-3 rounded-large [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover w-full"
            data-react-aria-pressable="true"
            tabindex="0"
            type="button"
          >
            <span />
            Return to Home
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OrderConfirmation Page > snapshots > renders order confirmation correctly 1`] = `
<h2>
  Unexpected Application Error!
</h2>
`;
